src/vellum/workflows/nodes/displayable/web_search_node/node.py

```diff
@@ -34,15 +42,93 @@ class Outputs(BaseOutputs):
         results: List[Dict[str, Any]]
 
     def _validate(self) -> None:
-        """Validate node inputs.
+        """Validate node inputs."""
+        if not self.query or not isinstance(self.query, str) or not self.query.strip():
+            raise NodeException(
+                "Query is required and must be a non-empty string", code=WorkflowErrorCode.INVALID_INPUTS
+            )
 
-        TODO: Add validation for query, api_key, and num_results parameters.
-        """
-        pass
+        if self.api_key is None:
+            raise NodeException("API key is required for SerpAPI", code=WorkflowErrorCode.INVALID_INPUTS)
+
+        if not isinstance(self.num_results, int) or self.num_results <= 0:
+            raise NodeException("num_results must be a positive integer", code=WorkflowErrorCode.INVALID_INPUTS)
 
     def run(self) -> Outputs:
-        """Run the WebSearchNode to perform web search.
+        """Run the WebSearchNode to perform web search via SerpAPI."""
+        self._validate()
 
-        TODO: Implement web search functionality with SerpAPI integration.
-        """
-        raise NotImplementedError("WebSearchNode functionality not yet implemented")
+        api_key_value = self.api_key.name if self.api_key and hasattr(self.api_key, "name") else str(self.api_key)
+
+        params = {
+            "q": self.query,
+            "api_key": api_key_value,
```

Comment:
[q] does serpApi have a way to pass in the api key via headers instead of a body param? if so this is [blocking]